src/p4p/test/system_client_server.py

```diff
@@ -0,0 +1,631 @@
+        'writeable': True,
+        'label': 'Greet',
+        'returns': {
+            'elements': {
+                'return': {
+                    'description': 'The manufactured greeting',
+                    'tags': ['widget:textupdate'],
+                    'writeable': False,
+                    'label': 'Return'
+                }
+            },
+            'required': ['return']
+        }
+    },
+    'error': {
+        'takes': empty_map_meta_dict,
+        'description': 'Raise an error',
+        'tags': [],
+        'writeable': True,
+        'label': 'Error',
+        'returns': empty_map_meta_dict
+    }
+}
+
+hello_expected = Value(hello_block_t, hello_dict)
+
+
+class Channel(object):
+    # TODO: Why is this needed?
+    Value = Value
+
+    def __init__(self, structure, field_name=None):
+        self.structure = structure
+        self.field_name = field_name
+
+    def get(self):
+        if self.field_name:
+            ret = self.structure[self.field_name]
+        else:
+            ret = self.structure
+        return ret
+
+    def monitor(self, updater):
+        # TODO: what is good here?
+        self.structure.attach(updater, self.field_name)
+
+    def rpc(self, response, request):
+        raise NotImplementedError()
+
+    def put(self, response, request):
+        raise NotImplementedError()
+
+
+class BlockChannel(Channel):
+    def __init__(self, structure, channels):
+        self.channels = channels
+        super(BlockChannel, self).__init__(structure)
+
+    def rpc(self, response, request):
+        # TODO: need the connection request object
+        method_name = connection_request["method"]
+        return self.channels[method_name].rpc(response, request)
+
+
+class Provider(object):
+    def __init__(self):
+        self.channels = {}
+
+    def add_block(self, name, structure, channel_classes):
+        # The channels we are going to be creating for this block
+        channels = {}
+        for k, _ in structure.items():
+            if k in channel_classes:
+                # We have been given some rpc/put behaviour
+                channel = channel_classes[k](structure, k)
+            else:
+                # Just use the generic channel
+                channel = Channel(structure, k)
+            channels[k] = channel
+        # One for the top level too
+        self.channels[name] = BlockChannel(structure, channels)
+        # Add the field channels in a dotted namespace
+        for k, channel in channels.items():
+            self.channels[name + "." + k] = channel
+
+    def testChannel(self, name):
+        return name in self.channels
+
+    def makeChannel(self, name, src):
+        return self.channels[name]
+
+
+def set_attr_ts(attr, value):
+    now = time.time()
+    attr.value = value
+    attr.timeStamp.secondsPastEpoch = int(now)
+    attr.timeStamp.nanoseconds = int(now % 1 / 1e-9)
+    # TODO: do we have to mark these as changed? And notify subscribers?
```

Comment:
Attribute assignment automatically marks them as changed.  Value objects have no notion of "subscriber" if you're thinking about ```PVField::postPut()```.